lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -948,6 +948,38 @@ public function set_attribute( $name, $value ) {
 			return;
 		}
 
+		/*
+		 * Verify that the attribute name is allowable. In WP_DEBUG
+		 * environments we want to crash to quickly alert developers
+		 * of typos and issues; but in production we don't want to
+		 * interrupt a normal page view, so we'll silently avoid
+		 * updating the attribute in those cases.
+		 *
+		 * Of note, we're disallowing more characters than are strictly
+		 * forbidden in HTML5. This is to prevent additional security
+		 * risks deeper in the WordPress and plugin stack. Specifically
+		 * we reject the less-than (<) and ampersand (&) characters.
+		 *
+		 * The use of a PCRE match allows us to look for specific Unicode
+		 * code points without writing a UTF-8 decoder. Whereas scanning
+		 * for one-byte characters is trivial, scanning for the longer
+		 * byte sequences would be more complicated, and this shouldn't
+		 * be in the hot path for execution so we can compromise on the
+		 * efficiency at this point.
+		 *
+		 * @see https://html.spec.whatwg.org/#attributes-2
+		 */
+		if ( preg_match(
+			'~[ "\'>&</=\x{00}-\x{1F}\x{FDD0}-\x{FDEF}\x{FFFE}\x{FFFF}\x{1FFFE}\x{1FFFF}\x{2FFFE}\x{2FFFF}\x{3FFFE}\x{3FFFF}\x{4FFFE}\x{4FFFF}\x{5FFFE}\x{5FFFF}\x{6FFFE}\x{6FFFF}\x{7FFFE}\x{7FFFF}\x{8FFFE}\x{8FFFF}\x{9FFFE}\x{9FFFF}\x{AFFFE}\x{AFFFF}\x{BFFFE}\x{BFFFF}\x{CFFFE}\x{CFFFF}\x{DFFFE}\x{DFFFF}\x{EFFFE}\x{EFFFF}\x{FFFFE}\x{FFFFF}\x{10FFFE}\x{10FFFF}]~Ssu',
```

Comment:
Is this any different than `strcspn`?

Reply:
Ah I see, you used ranges here: `\x{00}-\x{1F}`, `\x{FDD0}-\x{FDEF}`

Reply:
to be honest I assumed it wouldn't work with `strcspn` and didn't even try the Unicode code points. just now I confirmed on PHP 5.4.40 and it worked as expected - I'll update.

originally I had the control codes listed with `strcspn` but then moved to the `preg_match` for the ranges.

Reply:
@adamziel we might need to retain the `preg_match` because I think your test code is subtly wrong. as I suspected `strspn` and `strcspn` operate on normal PHP strings, which are byte-wise compared. what you're seeing is a rejection of strings containing any of the bytes `\xF3\x8F\xBF\xBE`, not those containing only that sequence.

```php
php > var_dump( strcspn( "test\u{CFFFE}test", "\u{CFFFE}" ) )
int(4)

php > var_dump( strcspn( "test\u{CFFFF}test", "\u{CFFFE}" ) )
int(4)

php > var_dump( strcspn( "test\x8ftest", "\u{CFFFE}" ) );
int(4)

php > var_dump( strcspn( "testtest", "\u{CFFFE}" ) );
int(8)
```

Reply:
Oh this is such a good spot 💯  Let's stick to `preg_match`, then.